### PR TITLE
Data handler

### PR DESCRIPTION
**Pull request by beamalsky**
## Overview

This PR implements a custom Fobi data handler to store collected survey data in PLDP-described models. 

### Notes

See comments in `fobi_form_handlers.py` for details on the parts of this that will need more work once we address #63. Study, Location, and Total fields are handled with static placeholder values for now.

Something to consider in `django-pldp`: now that SurveyComponent is more generic, it no longer makes sense to have a detail_level field that defaults to "basic". Maybe we should switch that to None?

Also, this needs some tests! @reginafcompton has kindly agreed to help me get started on that in the next few days, so I can pick it up on Monday.

## Testing Instructions

* Make sure to update requirements and pull in new migrations! Several things have changed in `django-pldp`
* Create a new survey locally. Check to make sure it has "Collect data" as a handler added automatically
Screen Shot 2019-03-27 at 4 58 36 PM

* Add at least one sample question to your new survey
* Run the survey! Submit an answer to your question
* Check the database. Do you see what you just submitted in the `pldp_survey`, `pldp_surveyrow`, and `pldp_surveycomponent` tables?

Connects #25

### CODE REVIEW

**Comment by beamalsky** [comment]
Moving this out of draft mode so that the Travis builds will run 👀

**Comment by reginafcompton** [comment]
For tests, I outlined a possible way forward in this PR: https://github.com/datamade/just-spaces/pull/68

I suggest that you write unit tests for the `run` function of `CollectDataPlugin`. I set-up the basic pattern for doing that, but I did not stub out actual tests. 

Where to begin? Before I write tests, I like to ask: what does this function do? In this case, the `run` function creates instances of SurveyRows and SurveyComponents. Then, I like to ask: what could go wrong? Why would the `run` function not behave as expected? Those sorts of questions can help you identify edge cases, improve your code, and write more robust tests. 

Let me know what you think. Happy to pair on this, too!

**Comment by beamalsky** [comment]
Hello @reginafcompton! I'll respond to your other feedback and very helpful test setup PR in a moment, but first let's make sure you can get this running locally. I forgot to make note of how to create a sample Location object—my apologies. I've gone ahead and made a data dump with the updated migrations, data handler, and sample metadata for the PLDP (Sample Study Area, Sample Area, Sample Location, and Sample Agency), which we can use on the staging site next time we're pushing to master. This involved deleting some outdated survey elements: `pldp_age_single`, `pldp_age_multiple`, `pldp_gender_single`, `pldp_gender_multiple`. These will need to be added back once we've completed [issue about 2 survey paths] and recreated those elements. 

## New Testing Instructions

- Drop the database: `dropdb just-spaces`
- Recreate the database: `pg_restore -C -j4 --no-owner just-spaces.dump | psql`
- Run the server: `python manage.py runserver`
- Login is the same as the staging site: testuser / 4SPUDO6f9Zm
- You should see the two example surveys: Porch Survey and Movies in Clark Park. Choose one of them, fill out some answers, and submit it.
- Check the database. Do you see what you just submitted in the pldp_survey, pldp_surveyrow, and pldp_surveycomponent tables?

I've also created a fixture—`pldp_sample`—that can reload a Sample Study Area, Sample Area, Sample Location, and Sample Agency in case we want that handy in the future. You shouldn't need to load that now, but for future reference the command is `python manage.py loaddata pldp_sample.json`

**Comment by beamalsky** [comment]
@reginafcompton I agree 100% that that tabbed view you screenshoted (Elements - Handlers - Properties - Service) should change, and not only because those terms won't really mean anything to users of the tool. I think Handlers and Service can definitely be hidden, but Properties depends on the outcome of #63 so I was waiting to do anything to that UI until we address our handling of Survey metadata.

File: fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        new_survey = Survey.objects.create(
+            study=study,
+            form_id=form_id,
+            location=location,
+            time_stop=time_stop
+        )
+
+        print("Survey submitted " + str(time_stop))
+
+        new_survey_row = SurveyRow.objects.create(
```
Comment (reginafcompton) [inline fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: Should the creation of a SurveyRow be nested under the iteration over `form_elements` (lone 53 ff.)? If I understand correctly, we'll create a survey row (with the number people counted by the question) with each survey component? 

I also wonder if the `total` default for intercept surveys should be 1. Can you say more about your reasoning?

File: fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        time_stop = datetime.datetime.now(tz=get_current_timezone())
+
+        new_survey = Survey.objects.create(
+            study=study,
+            form_id=form_id,
+            location=location,
+            time_stop=time_stop
+        )
+
+        print("Survey submitted " + str(time_stop))
```
Comment (reginafcompton) [inline fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: Let's discard this print statement before merging the PR.

File: fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py (hunk 1 above)
Comment (beamalsky) [reply fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: Hm, excellent question! The separate SurveyRow and SurveyComponent models come from Eric's original implementation of `django-pldp`, and I've been following that: the assumption is that each submitted survey (not each question in it) will have a required field with the total number of people being counted in that survey. If it's an intercept survey, with just one respondent, the total number of people counted would be 1. This will be useful in later analysis, when we want to make claims like "x% of people were riding bicycles."

Now that we have more example PLDP documents in Gehl's repo, I can check those assumptions! It seems this understanding is supported in their suggested [data structure](https://github.com/gehl-institute/pldp/blob/master/Public%20Life%20Data%20Protocol/PLDP_1.0%20Data%20Sheet%20Template%20(Single%20Table)). 

However, not all of their sample data (like [this one](https://github.com/gehl-institute/pldp/blob/master/Sample%20Data/sf-actscan.csv)) includes that total field at all. Maybe the total default should be None, and it should be assigned to 1 for intercept surveys specifically?

Comment (beamalsky) [reply fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: > The expectation is that the total number of people counted in observational surveys does not change from question-to-question.

Yes!

A reason to save 1 as a `total` for intercept surveys is that this would make it possible to compare across multiple submissions of the same survey form. For example, noting that out of 50 surveys, 40 of them responded "yes" to "Are you riding a bicycle?"

This is a question of how we're handling analysis, and it could also be derived later from noting that the survey is intercept. However, it makes more sense to me to save that in the `total` field while we're submitting the form and make a distinction between a survey regarding a single person and a survey regarding an unknown number of people.

**Comment by beamalsky** [comment]
@reginafcompton I have a working test here that creates a sample survey, surveyrow, and surveycomponent, then checks that they're all existent and linked. 

I struggled to actually populate the sample survey with submitted data. Here is what I want it to have: 

`form.cleaned_data['c75e27cf-4d8f-49dc-bb15-da8137dac247'] = 5`

File: tests/test_data_handling.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from fobi.models import FormEntry
+
+from pldp.models import Survey, SurveyRow, SurveyComponent
+
+@pytest.mark.django_db
+def test_data_handler(sample_form_entry, sample_location, sample_study, sample_form_element, mocker):
+    request = mocker.MagicMock(spec=requests.Response)
+    form = assemble_form_class(form_entry=sample_form_entry)
+
+    # form.cleaned_data['c75e27cf-4d8f-49dc-bb15-da8137dac247'] = 5
```
Comment (reginafcompton) [inline tests/test_data_handling.py]: All right. This is tricky. I looked at [`django-fobi` for inspiration.](https://github.com/barseghyanartur/django-fobi/blob/070ee3239cc4df3f5e841ee8649d8c26a10f007e/src/fobi/tests/test_dynamic_forms.py#L35)

```
# Instantiate an instance of the `assemble_form_class`.
form_class = assemble_form_class(form_entry=sample_form_entry)

# Pass data to it
form = form_class(data={'c75e27cf-4d8f-49dc-bb15-da8137dac247': 5})

assert form.is_valid() == True
```

I have a couple questions about this:
* I do not understand the GUIDs as form fields. Do you know what fobi has going on here?
* What is your motivation for creating a `DynamicForm` with data? Your `run` function does not appear to do anything with it directly, right?

File: tests/test_data_handling.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    # A valid CollectDataPlugin requires a DynamicForm, a FormEntry, and a post request
+    plugin = CollectDataPlugin()
+    plugin.run(form=form,
+               form_entry=sample_form_entry,
+               request=request)
+
+    # Check that a survey has been created with a linked row and component
+    survey = Survey.objects.first()
+    row = SurveyRow.objects.filter(survey=survey)
+    component = SurveyComponent.objects.filter(row=row[0])
```
Comment (reginafcompton) [inline tests/test_data_handling.py]: Can you save `row[0]` in a variable? That would help future maintainers (and future you) understand what you're filtering on.

File: tests/test_data_handling.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    assert (len(row), len(component)) == (1, 1)
```
Comment (reginafcompton) [inline tests/test_data_handling.py]: I think this is a good assertion to make. But I'd also like you to make some assertions about the content of the newly created entities. Does the SurveyRow and SurveyComponent have the data you expect, given your fixtures?

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        call_command('initialize_pldp')
+
+
+@pytest.fixture
+@pytest.mark.django_db
+def sample_user(db):
+    sample_user = JustSpacesUser.objects.create(
+        username='sampleuser'
+    )
+    sample_user.save()
```
[the remark this answers is not in the document]
Comment (beamalsky) [reply tests/conftest.py]: eep! thanks

File: tests/test_data_handling.py (hunk 3 above)
Comment (beamalsky) [reply tests/test_data_handling.py]: - Fobi has a 'name' field on each form element, which it uses to connect saved submitted data with other form data (eg plugin_uid, position, etc.). It's not the way I would have set up the package, necessarily, but changing it would mean doing some pretty major surgery to Fobi's base functions. It's essentially an additional ID for each form element, as I understand it.
- The run function saves submitted data! It should end up in the `saved_data` column of SurveyComponent. See: https://github.com/datamade/just-spaces/blob/data-handler/fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py#L67

File: fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py (hunk 1 above)
Comment (beamalsky) [reply fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: @reginafcompton what do you think about this?

File: tests/test_data_handling.py (hunk 4 above)
Comment (beamalsky) [reply tests/test_data_handling.py]: yep!

File: tests/test_data_handling.py (hunk 5 above)
Comment (beamalsky) [reply tests/test_data_handling.py]: I added a lot more! This now checks for all fields that should be saved. Is this too much? It seems like too much for a single test to handle but I'm not sure how to break it up.

File: tests/test_data_handling.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    # find its linked components
+    row = rows[0]
+    components = SurveyComponent.objects.filter(row=row)
+    component = components[0]
+
+    # Check that the handler has created a single row and linked component
+    assert len(rows) == 1
+    assert len(components) == 1
+
+    # Check survey has the correct values
```
Comment (reginafcompton) [inline tests/test_data_handling.py]: Most of this code does not need the sort of obvious commenting you have here. 

If you feel like the code needs some explanation, you should place that in a docstring at the top of the text, i.e., right after line 12. It's not a requirement of this PR. Just something to think about.

File: fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py (hunk 1 above)
Comment (reginafcompton) [reply fobi_custom/plugins/form_handlers/collect_data/fobi_form_handlers.py]: Okay, I think 1 as default for intercept surveys makes sense, given your example above. Let's plan for that. And, as you note, this particular detail may change depending on how the tool handles analysis.
